File: src/main/java/data/parser/me7log/Me7LogParser.java
package data.parser.me7log;

import data.contract.Me7LogFileContract;
import io.reactivex.annotations.NonNull;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVRecord;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Me7LogParser {

    public enum LogType {
        OPEN_LOOP,
        CLOSED_LOOP,
        LDRPID,
        KFVPDKSD
    }

    private int timeColumnIndex = -1;
    private int rpmColumnIndex = -1;
    private int stftColumnIndex = -1;
    private int ltftColumnIndex = -1;
    private int mafVoltageIndex = -1;
    private int mafGramsPerSecondIndex = -1;
    private int throttlePlateAngleIndex = -1;
    private int lambdaControlActiveIndex = -1;
    private int requestedLambdaIndex = -1;
    private int fuelInjectorOnTimeIndex = -1;
    private int engineLoadIndex = -1;
    private int wastegateDutyCycleIndex = -1;
    private int barometricPressureIndex = -1;
    private int absoluteBoostPressureActualIndex = -1;
    private int selectedGearIndex = -1;

    public interface ProgressCallback {
        void onProgress(int value, int max);
    }

    public Map<Me7LogFileContract.Header, List<Double>> parseLogDirectory(LogType logType, File directory, ProgressCallback callback) {
        Map<Me7LogFileContract.Header, List<Double>> map = generateMap(logType);
        File[] files = directory.listFiles();
        int numFiles = files.length;
        int count = 0;

        for (File file : files) {
            parse(file, logType, map);
            callback.onProgress(count++, numFiles);
        }

        return map;
    }

    @NonNull
    public Map<Me7LogFileContract.Header, List<Double>> parseLogFile(LogType logType, File file) {
        Map<Me7LogFileContract.Header, List<Double>> map = generateMap(logType);

        parse(file, logType, map);

        return map;
    }

    private void parse(File file, LogType logType, Map<Me7LogFileContract.Header, List<Double>> map) {
        resetIndices();
        try {
            boolean headersFound = false;
            Reader in = new FileReader(file);
            Iterable<CSVRecord> records = CSVFormat.RFC4180.parse(in);
            for (CSVRecord record : records) {
                if (record.size() > 0) {
                    String string = record.get(0);
                    if (string.contains("Log started at:")) {
                        String[] split = string.split(" ");
                        String timestamp = split[5];
                        split = timestamp.split(":");
                        double minuteSeconds = Double.parseDouble(split[1]) * 60;
                        double secondsSeconds = Double.parseDouble(split[2]);

                        double startTime = (minuteSeconds + secondsSeconds);

                        map.get(Me7LogFileContract.Header.START_TIME_HEADER).add(startTime);
                    }
                }

                for (int i = 0; i < record.size(); i++) {
                    if (Me7LogFileContract.Header.TIME_COLUMN_HEADER.getHeader().equals(record.get(i).trim())) {
                        timeColumnIndex = i;
                    } else if (Me7LogFileContract.Header.RPM_COLUMN_HEADER.getHeader().equals(record.get(i).trim())) {
                        rpmColumnIndex = i;
                    } else if (Me7LogFileContract.Header.STFT_COLUMN_HEADER.getHeader().equals(record.get(i).trim())) {
                        stftColumnIndex = i;
                    } else if (Me7LogFileContract.Header.LTFT_COLUMN_HEADER.getHeader().equals(record.get(i).trim())) {
                        ltftColumnIndex = i;
                    } else if (Me7LogFileContract.Header.MAF_VOLTAGE_HEADER.getHeader().equals(record.get(i).trim())) {
                        mafVoltageIndex = i;
                    } else if (Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER.getHeader().equals(record.get(i).trim())) {
                        mafGramsPerSecondIndex = i;
                    } else if (Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER.getHeader().equals(record.get(i).trim())) {
                        throttlePlateAngleIndex = i;
                    } else if (Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER.getHeader().equals(record.get(i).trim())) {
                        lambdaControlActiveIndex = i;
                    } else if (Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER.getHeader().equals(record.get(i).trim())) {
                        requestedLambdaIndex = i;
                    } else if (Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER.getHeader().equals(record.get(i).trim())) {
                        fuelInjectorOnTimeIndex = i;
                    } else if (Me7LogFileContract.Header.ENGINE_LOAD_HEADER.getHeader().equals(record.get(i).trim())) {
                        engineLoadIndex = i;
                    } else if (Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER.getHeader().equals(record.get(i).trim())) {
                        absoluteBoostPressureActualIndex = i;
                    } else if (Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER.getHeader().equals(record.get(i).trim())) {
                        barometricPressureIndex = i;
                    } else if (Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER.getHeader().equals(record.get(i).trim())) {
                        wastegateDutyCycleIndex = i;
                    } else if (Me7LogFileContract.Header.SELECTED_GEAR_HEADER.getHeader().equals(record.get(i).trim())) {
                        selectedGearIndex = i;
                    }

                    headersFound = headersFound(logType);

                    if (headersFound) {
                        break;
                    }
                }

                if (headersFound) {
                    break;
                }
            }

            if (headersFound) {
                for (CSVRecord record : records) {
                    try {
                        if (logType == LogType.CLOSED_LOOP || logType == LogType.OPEN_LOOP) {
                            double time = Double.parseDouble(record.get(timeColumnIndex));
                            double rpm = Double.parseDouble(record.get(rpmColumnIndex));
                            double stft = Double.parseDouble(record.get(stftColumnIndex));
                            double ltft = Double.parseDouble(record.get(ltftColumnIndex));
                            double mafVoltage = Double.parseDouble(record.get(mafVoltageIndex));
                            double throttlePlateAngle = Double.parseDouble(record.get(throttlePlateAngleIndex));
                            double lambdaControlActive = Double.parseDouble(record.get(lambdaControlActiveIndex));
                            double engineLoad = Double.parseDouble(record.get(engineLoadIndex));

                            if (logType == LogType.OPEN_LOOP) {
                                double mafGsec = Double.parseDouble(record.get(mafGramsPerSecondIndex));
                                double requestedLambda = Double.parseDouble(record.get(requestedLambdaIndex));
                                double fuelInjectorOnTime = Double.parseDouble(record.get(fuelInjectorOnTimeIndex));

                                map.get(Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER).add(mafGsec);
                                map.get(Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER).add(requestedLambda);
                                map.get(Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER).add(fuelInjectorOnTime);
                            }

                            map.get(Me7LogFileContract.Header.TIME_COLUMN_HEADER).add(time);
                            map.get(Me7LogFileContract.Header.RPM_COLUMN_HEADER).add(rpm);
                            map.get(Me7LogFileContract.Header.STFT_COLUMN_HEADER).add(stft);
                            map.get(Me7LogFileContract.Header.LTFT_COLUMN_HEADER).add(ltft);
                            map.get(Me7LogFileContract.Header.MAF_VOLTAGE_HEADER).add(mafVoltage);
                            map.get(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER).add(throttlePlateAngle);
                            map.get(Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER).add(lambdaControlActive);
                            map.get(Me7LogFileContract.Header.ENGINE_LOAD_HEADER).add(engineLoad);
                        } else if (logType == LogType.LDRPID) {
                            double time = Double.parseDouble(record.get(timeColumnIndex));
                            double rpm = Double.parseDouble(record.get(rpmColumnIndex));
                            double throttlePlateAngle = Double.parseDouble(record.get(throttlePlateAngleIndex));
                            double barometricPressure = Double.parseDouble(record.get(barometricPressureIndex));
                            double wastegateDutyCycle = Double.parseDouble(record.get(wastegateDutyCycleIndex));
                            double absoluteBoostPressure = Double.parseDouble(record.get(absoluteBoostPressureActualIndex));
                            double selectedGear = Double.parseDouble(record.get(selectedGearIndex));

                            map.get(Me7LogFileContract.Header.TIME_COLUMN_HEADER).add(time);
                            map.get(Me7LogFileContract.Header.RPM_COLUMN_HEADER).add(rpm);
                            map.get(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER).add(throttlePlateAngle);
                            map.get(Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER).add(barometricPressure);
                            map.get(Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER).add(wastegateDutyCycle);
                            map.get(Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER).add(absoluteBoostPressure);
                            map.get(Me7LogFileContract.Header.SELECTED_GEAR_HEADER).add(selectedGear);
                        } else if (logType == LogType.KFVPDKSD) {
                            double time = Double.parseDouble(record.get(timeColumnIndex));
                            double rpm = Double.parseDouble(record.get(rpmColumnIndex));
                            double throttlePlateAngle = Double.parseDouble(record.get(throttlePlateAngleIndex));
                            double barometricPressure = Double.parseDouble(record.get(barometricPressureIndex));
                            double absoluteBoostPressure = Double.parseDouble(record.get(absoluteBoostPressureActualIndex));

                            map.get(Me7LogFileContract.Header.TIME_COLUMN_HEADER).add(time);
                            map.get(Me7LogFileContract.Header.RPM_COLUMN_HEADER).add(rpm);
                            map.get(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER).add(throttlePlateAngle);
                            map.get(Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER).add(barometricPressure);
                            map.get(Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER).add(absoluteBoostPressure);
                        }
                    } catch (NumberFormatException | ArrayIndexOutOfBoundsException e) {
                        e.printStackTrace();
                    }
                }
            }
        } catch (IOException | ArrayIndexOutOfBoundsException e) {
            e.printStackTrace();
        }

        int size = -1;
        for (Me7LogFileContract.Header key : map.keySet()) {
            if (!key.equals(Me7LogFileContract.Header.START_TIME_HEADER) && size == -1) {
                size = map.get(key).size();
            }

            if (!key.equals(Me7LogFileContract.Header.START_TIME_HEADER) && map.get(key).size() != size) {
                throw new RuntimeException("Data is not square! Got: " + map.get(key).size() + " Expected: " + size);
            }
        }
    }

    private void resetIndices() {
        timeColumnIndex = -1;
        rpmColumnIndex = -1;
        stftColumnIndex = -1;
        ltftColumnIndex = -1;
        mafVoltageIndex = -1;
        mafGramsPerSecondIndex = -1;
        throttlePlateAngleIndex = -1;
        lambdaControlActiveIndex = -1;
        requestedLambdaIndex = -1;
        fuelInjectorOnTimeIndex = -1;
        engineLoadIndex = -1;
        wastegateDutyCycleIndex = -1;
        barometricPressureIndex = -1;
        absoluteBoostPressureActualIndex = -1;
        selectedGearIndex = -1;
    }

    private boolean headersFound(LogType logType) {
        if (logType == LogType.OPEN_LOOP) {
            return timeColumnIndex != -1 && rpmColumnIndex != -1 && stftColumnIndex != -1 && ltftColumnIndex != -1 && mafVoltageIndex != -1 && mafGramsPerSecondIndex != -1 && throttlePlateAngleIndex != -1 && lambdaControlActiveIndex != -1 && requestedLambdaIndex != -1 && fuelInjectorOnTimeIndex != -1;
        } else if (logType == LogType.CLOSED_LOOP) {
            // nmot, fr_w, fra_2, uhfm_w, wdkba, B_lr, rl_w
            return timeColumnIndex != -1 && rpmColumnIndex != -1 && stftColumnIndex != -1 && ltftColumnIndex != -1 && mafVoltageIndex != -1 && throttlePlateAngleIndex != -1 && lambdaControlActiveIndex != -1 && engineLoadIndex != -1;
        } else if (logType == LogType.LDRPID) {
            return timeColumnIndex != -1 && rpmColumnIndex != -1 && throttlePlateAngleIndex != -1 && wastegateDutyCycleIndex != -1 && barometricPressureIndex != -1 && absoluteBoostPressureActualIndex != -1 && selectedGearIndex != -1;
        } else if (logType == LogType.KFVPDKSD) {
            return timeColumnIndex != -1 && rpmColumnIndex != -1 && throttlePlateAngleIndex != -1 && barometricPressureIndex != -1 && absoluteBoostPressureActualIndex != -1;
        }

        return false;
    }

    private Map<Me7LogFileContract.Header, List<Double>> generateMap(LogType logType) {
        Map<Me7LogFileContract.Header, List<Double>> map = new HashMap<>();

        map.put(Me7LogFileContract.Header.START_TIME_HEADER, new ArrayList<>());

        if (logType == LogType.CLOSED_LOOP || logType == LogType.OPEN_LOOP) {
            map.put(Me7LogFileContract.Header.TIME_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.RPM_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.STFT_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.LTFT_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.MAF_VOLTAGE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.ENGINE_LOAD_HEADER, new ArrayList<>());

            if (logType == LogType.OPEN_LOOP) {
                map.put(Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER, new ArrayList<>());
                map.put(Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER, new ArrayList<>());
                map.put(Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER, new ArrayList<>());
            }
        } else if (logType == LogType.LDRPID) {
            map.put(Me7LogFileContract.Header.TIME_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.RPM_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.SELECTED_GEAR_HEADER, new ArrayList<>());
        } else if (logType == LogType.KFVPDKSD) {
            map.put(Me7LogFileContract.Header.TIME_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.RPM_COLUMN_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER, new ArrayList<>());
            map.put(Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER, new ArrayList<>());
        }

        return map;
    }
}

File: src/main/kotlin/data/parser/me7log/Me7LogParser.kt
package data.parser.me7log

import data.contract.Me7LogFileContract
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVRecord
import java.io.File
import java.io.FileReader

class Me7LogParser {

    enum class LogType {
        OPEN_LOOP,
        CLOSED_LOOP,
        LDRPID,
        KFVPDKSD
    }

    fun interface ProgressCallback {
        fun onProgress(value: Int, max: Int)
    }

    private var timeColumnIndex = -1
    private var rpmColumnIndex = -1
    private var stftColumnIndex = -1
    private var ltftColumnIndex = -1
    private var mafVoltageIndex = -1
    private var mafGramsPerSecondIndex = -1
    private var throttlePlateAngleIndex = -1
    private var lambdaControlActiveIndex = -1
    private var requestedLambdaIndex = -1
    private var fuelInjectorOnTimeIndex = -1
    private var engineLoadIndex = -1
    private var wastegateDutyCycleIndex = -1
    private var barometricPressureIndex = -1
    private var absoluteBoostPressureActualIndex = -1
    private var selectedGearIndex = -1
    private var wideBandO2Index = -1

    fun parseLogDirectory(logType: LogType, directory: File, callback: ProgressCallback): Map<Me7LogFileContract.Header, List<Double>> {
        val map = generateMap(logType)
        val files = directory.listFiles() ?: return map
        val numFiles = files.size
        var count = 0

        for (file in files) {
            parse(file, logType, map)
            callback.onProgress(count++, numFiles)
        }

        return map
    }

    fun parseLogFile(logType: LogType, file: File): Map<Me7LogFileContract.Header, List<Double>> {
        val map = generateMap(logType)
        parse(file, logType, map)
        return map
    }

    private fun parse(file: File, logType: LogType, map: Map<Me7LogFileContract.Header, MutableList<Double>>) {
        resetIndices()
        try {
            FileReader(file).use { reader ->
                var headersFound = false
                val records = CSVFormat.RFC4180.parse(reader)
                val iterator = records.iterator()

                while (iterator.hasNext()) {
                    val record = iterator.next()
                    if (record.size() > 0) {
                        val string = record.get(0)
                        if (string.contains("Log started at:")) {
                            val split = string.split(" ")
                            val timestamp = split[5]
                            val timeParts = timestamp.split(":")
                            val minuteSeconds = timeParts[1].toDouble() * 60
                            val secondsSeconds = timeParts[2].toDouble()
                            val startTime = minuteSeconds + secondsSeconds
                            map[Me7LogFileContract.Header.START_TIME_HEADER]!!.add(startTime)
                        }
                    }

                    for (i in 0 until record.size()) {
                        val trimmed = record.get(i).trim()
                        when {
                            Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER.header == trimmed -> timeColumnIndex = i
                            Me7LogFileContract.Header.RPM_COLUMN_HEADER.header == trimmed -> rpmColumnIndex = i
                            Me7LogFileContract.Header.STFT_COLUMN_HEADER.header == trimmed -> stftColumnIndex = i
                            Me7LogFileContract.Header.LTFT_COLUMN_HEADER.header == trimmed -> ltftColumnIndex = i
                            Me7LogFileContract.Header.MAF_VOLTAGE_HEADER.header == trimmed -> mafVoltageIndex = i
                            Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER.header == trimmed -> mafGramsPerSecondIndex = i
                            Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER.header == trimmed -> throttlePlateAngleIndex = i
                            Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER.header == trimmed -> lambdaControlActiveIndex = i
                            Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER.header == trimmed -> requestedLambdaIndex = i
                            Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER.header == trimmed -> fuelInjectorOnTimeIndex = i
                            Me7LogFileContract.Header.ENGINE_LOAD_HEADER.header == trimmed -> engineLoadIndex = i
                            Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER.header == trimmed -> absoluteBoostPressureActualIndex = i
                            Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER.header == trimmed -> barometricPressureIndex = i
                            Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER.header == trimmed -> wastegateDutyCycleIndex = i
                            Me7LogFileContract.Header.SELECTED_GEAR_HEADER.header == trimmed -> selectedGearIndex = i
                            Me7LogFileContract.Header.WIDE_BAND_O2_HEADER.header == trimmed -> wideBandO2Index = i
                        }

                        headersFound = headersFound(logType)
                        if (headersFound) break
                    }

                    if (headersFound) break
                }

                if (headersFound) {
                    while (iterator.hasNext()) {
                        val record = iterator.next()
                        try {
                            when (logType) {
                                LogType.CLOSED_LOOP, LogType.OPEN_LOOP -> {
                                    val time = record.get(timeColumnIndex).toDouble()
                                    val rpm = record.get(rpmColumnIndex).toDouble()
                                    val stft = record.get(stftColumnIndex).toDouble()
                                    val ltft = record.get(ltftColumnIndex).toDouble()
                                    val mafVoltage = record.get(mafVoltageIndex).toDouble()
                                    val throttlePlateAngle = record.get(throttlePlateAngleIndex).toDouble()
                                    val lambdaControlActive = record.get(lambdaControlActiveIndex).toDouble()
                                    val engineLoad = record.get(engineLoadIndex).toDouble()

                                    if (logType == LogType.OPEN_LOOP) {
                                        val mafGsec = record.get(mafGramsPerSecondIndex).toDouble()
                                        val requestedLambda = record.get(requestedLambdaIndex).toDouble()
                                        val fuelInjectorOnTime = record.get(fuelInjectorOnTimeIndex).toDouble()

                                        map[Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER]!!.add(mafGsec)
                                        map[Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER]!!.add(requestedLambda)
                                        map[Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER]!!.add(fuelInjectorOnTime)

                                        if (wideBandO2Index != -1) {
                                            val wideBandO2 = record.get(wideBandO2Index).toDouble()
                                            map[Me7LogFileContract.Header.WIDE_BAND_O2_HEADER]!!.add(wideBandO2)
                                        }
                                    }

                                    map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER]!!.add(time)
                                    map[Me7LogFileContract.Header.RPM_COLUMN_HEADER]!!.add(rpm)
                                    map[Me7LogFileContract.Header.STFT_COLUMN_HEADER]!!.add(stft)
                                    map[Me7LogFileContract.Header.LTFT_COLUMN_HEADER]!!.add(ltft)
                                    map[Me7LogFileContract.Header.MAF_VOLTAGE_HEADER]!!.add(mafVoltage)
                                    map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER]!!.add(throttlePlateAngle)
                                    map[Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER]!!.add(lambdaControlActive)
                                    map[Me7LogFileContract.Header.ENGINE_LOAD_HEADER]!!.add(engineLoad)
                                }
                                LogType.LDRPID -> {
                                    map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER]!!.add(record.get(timeColumnIndex).toDouble())
                                    map[Me7LogFileContract.Header.RPM_COLUMN_HEADER]!!.add(record.get(rpmColumnIndex).toDouble())
                                    map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER]!!.add(record.get(throttlePlateAngleIndex).toDouble())
                                    map[Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER]!!.add(record.get(barometricPressureIndex).toDouble())
                                    map[Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER]!!.add(record.get(wastegateDutyCycleIndex).toDouble())
                                    map[Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER]!!.add(record.get(absoluteBoostPressureActualIndex).toDouble())
                                    map[Me7LogFileContract.Header.SELECTED_GEAR_HEADER]!!.add(record.get(selectedGearIndex).toDouble())
                                }
                                LogType.KFVPDKSD -> {
                                    map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER]!!.add(record.get(timeColumnIndex).toDouble())
                                    map[Me7LogFileContract.Header.RPM_COLUMN_HEADER]!!.add(record.get(rpmColumnIndex).toDouble())
                                    map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER]!!.add(record.get(throttlePlateAngleIndex).toDouble())
                                    map[Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER]!!.add(record.get(barometricPressureIndex).toDouble())
                                    map[Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER]!!.add(record.get(absoluteBoostPressureActualIndex).toDouble())
                                }
                            }
                        } catch (_: NumberFormatException) {
                        } catch (_: ArrayIndexOutOfBoundsException) {
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // Validate data is square
        var size = -1
        for ((key, value) in map) {
            if (key != Me7LogFileContract.Header.START_TIME_HEADER && size == -1) {
                size = value.size
            }
            if (key != Me7LogFileContract.Header.START_TIME_HEADER && value.size != size) {
                // Wideband is optional
                if (key != Me7LogFileContract.Header.WIDE_BAND_O2_HEADER) {
                    throw RuntimeException("Data is not square! Got: ${value.size} Expected: $size")
                }
            }
        }
    }

    private fun resetIndices() {
        timeColumnIndex = -1
        rpmColumnIndex = -1
        stftColumnIndex = -1
        ltftColumnIndex = -1
        mafVoltageIndex = -1
        mafGramsPerSecondIndex = -1
        throttlePlateAngleIndex = -1
        lambdaControlActiveIndex = -1
        requestedLambdaIndex = -1
        fuelInjectorOnTimeIndex = -1
        engineLoadIndex = -1
        wastegateDutyCycleIndex = -1
        barometricPressureIndex = -1
        absoluteBoostPressureActualIndex = -1
        selectedGearIndex = -1
        wideBandO2Index = -1
    }

    private fun headersFound(logType: LogType): Boolean = when (logType) {
        LogType.OPEN_LOOP -> timeColumnIndex != -1 && rpmColumnIndex != -1 && stftColumnIndex != -1 && ltftColumnIndex != -1 && mafVoltageIndex != -1 && mafGramsPerSecondIndex != -1 && throttlePlateAngleIndex != -1 && lambdaControlActiveIndex != -1 && requestedLambdaIndex != -1 && fuelInjectorOnTimeIndex != -1
        LogType.CLOSED_LOOP -> timeColumnIndex != -1 && rpmColumnIndex != -1 && stftColumnIndex != -1 && ltftColumnIndex != -1 && mafVoltageIndex != -1 && throttlePlateAngleIndex != -1 && lambdaControlActiveIndex != -1 && engineLoadIndex != -1
        LogType.LDRPID -> timeColumnIndex != -1 && rpmColumnIndex != -1 && throttlePlateAngleIndex != -1 && wastegateDutyCycleIndex != -1 && barometricPressureIndex != -1 && absoluteBoostPressureActualIndex != -1 && selectedGearIndex != -1
        LogType.KFVPDKSD -> timeColumnIndex != -1 && rpmColumnIndex != -1 && throttlePlateAngleIndex != -1 && barometricPressureIndex != -1 && absoluteBoostPressureActualIndex != -1
    }

    private fun generateMap(logType: LogType): Map<Me7LogFileContract.Header, MutableList<Double>> {
        val map = mutableMapOf<Me7LogFileContract.Header, MutableList<Double>>()
        map[Me7LogFileContract.Header.START_TIME_HEADER] = mutableListOf()

        when (logType) {
            LogType.CLOSED_LOOP, LogType.OPEN_LOOP -> {
                map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.RPM_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.STFT_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.LTFT_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.MAF_VOLTAGE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.LAMBDA_CONTROL_ACTIVE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.ENGINE_LOAD_HEADER] = mutableListOf()

                if (logType == LogType.OPEN_LOOP) {
                    map[Me7LogFileContract.Header.MAF_GRAMS_PER_SECOND_HEADER] = mutableListOf()
                    map[Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER] = mutableListOf()
                    map[Me7LogFileContract.Header.FUEL_INJECTOR_ON_TIME_HEADER] = mutableListOf()
                    map[Me7LogFileContract.Header.WIDE_BAND_O2_HEADER] = mutableListOf()
                }
            }
            LogType.LDRPID -> {
                map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.RPM_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.WASTEGATE_DUTY_CYCLE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.SELECTED_GEAR_HEADER] = mutableListOf()
            }
            LogType.KFVPDKSD -> {
                map[Me7LogFileContract.Header.TIME_STAMP_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.RPM_COLUMN_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.THROTTLE_PLATE_ANGLE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.BAROMETRIC_PRESSURE_HEADER] = mutableListOf()
                map[Me7LogFileContract.Header.ABSOLUTE_BOOST_PRESSURE_ACTUAL_HEADER] = mutableListOf()
            }
        }

        return map
    }
}
